File: fractran/parse.py
from pathlib import Path
import re

import numpy as np

from base import Rule, Program
from primes import prime_factor


def parse_fractions(prog_str: str) -> Program:
  rule_strs = prog_str.strip().removeprefix("[").removesuffix("]").split(",")
  fracs = []
  num_reg = 0
  for rule_str in rule_strs:
    p, q = rule_str.split("/")
    top = prime_factor(int(p))
    bot = prime_factor(int(q))
    fracs.append((top, bot))
    num_reg = max(num_reg, len(top), len(bot))
  prog = []
  for (top, bot) in fracs:
    prog.append(Rule.from_ratio(top, bot, num_reg))
  return Program(prog)

def parse_vectors(prog_str: str) -> Program:
  matches = re.findall("\[ *([-+0-9 ]*?) *\]", prog_str)
  rows = []
  max_width = 0
  for match in matches:
    if match:
      try:
        row = np.fromstring(match, dtype=int, sep=" ")
      except:
        print("Error parsing row:", match)
        raise
      max_width = max(max_width, row.size)
      rows.append(row)
  prog = []
  for row in rows:
    row.resize(max_width)
    prog.append(Rule(row))
  return Program(prog)

def parse_program(prog_str: str) -> Program:
  if "/" in prog_str:
    return parse_fractions(prog_str)
  else:
    return parse_vectors(prog_str)

def read_file(filename_record: str) -> str:
  if ":" in filename_record:
    filename, record_num = filename_record.split(":")
    filename = Path(filename)
    record_num = int(record_num)
  else:
    filename = Path(filename_record)
    record_num = None
  if not filename.exists():
    raise FileExistsError(filename)
  with open(filename, "r") as f:
    if record_num is not None:
      # TODO: This assumes 1 record per line, which is not "currently" true for vec format.
      return f.readlines()[record_num]
    else:
      return f.read()

def load_program(prog_or_filename: str) -> Program:
  if not prog_or_filename.startswith("["):
    prog_str = read_file(prog_or_filename)
  else:
    prog_str = str(prog_or_filename)

  return parse_program(prog_str)

def enum_programs(filename: Path):
  with open(filename, "r") as f:
    for line in f:
      yield parse_program(line)

File: fractran/rust/src/parse.rs
use crate::program::{Instr, Program};
use primal::Primes;
use prime_factorization::Factorization;
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead};

// Parse a Fractran program and convert into vector form.
pub fn parse_program(program_str: &str) -> Program {
    // 1. Clean and split string
    let mut inner = program_str;
    if let Some(start) = inner.find('[') {
        if let Some(end) = inner.rfind(']') {
            inner = &inner[start + 1..end];
        } else {
            panic!("Program string missing closing ']'");
        }
    } else {
        panic!("Program string missing opening '['");
    }

    // 2. Parse fractions and find max prime
    let mut instrs_fractions: Vec<(u128, u128)> = Vec::new();
    let mut max_prime_found: u128 = 2;

    let normalized = inner.replace(',', " ");
    for part in normalized.split_whitespace() {
        if !part.contains('/') {
            panic!("Invalid fraction format (missing '/'): {}", part);
        }

        let frac: Vec<&str> = part.split('/').collect();
        if frac.len() != 2 {
            panic!("Invalid fraction format: {}", part);
        }

        let num: u128 = frac[0].parse().unwrap_or_else(|_| panic!("Invalid numerator in {}", part));
        let den: u128 = frac[1].parse().unwrap_or_else(|_| panic!("Invalid denominator in {}", part));
        instrs_fractions.push((num, den));

        // Check factors to find the largest prime needed for dimensions
        // We iterate the factors to find the max
        let num_factors = Factorization::run(num);
        let den_factors = Factorization::run(den);

        if let Some(&max) = num_factors.factors.iter().max() {
            if max > max_prime_found {
                max_prime_found = max;
            }
        }
        if let Some(&max) = den_factors.factors.iter().max() {
            if max > max_prime_found {
                max_prime_found = max;
            }
        }
    }

    // 3. Generate prime map (Prime -> Index) using `primal` crate
    // We map standard primes 2->0, 3->1, 5->2... up to max_prime_found
    let mut prime_map = HashMap::new();
    let mut dims = 0;

    // Primes::all() returns an iterator of usize. We cast to u128.
    for (i, p) in Primes::all().enumerate() {
        let p_u128 = p as u128;
        prime_map.insert(p_u128, i);
        dims = i + 1;
        if p_u128 >= max_prime_found {
            break;
        }
    }

    // 4. Build Matrix
    let mut instrs: Vec<Instr> = Vec::new();

    for (num, den) in instrs_fractions.iter() {
        let mut instr = vec![0; dims];
        // Handle Numerator (Additions)
        let num_factors = Factorization::run(*num);
        for p in num_factors.factors {
            if let Some(&col) = prime_map.get(&p) {
                instr[col] += 1;
            }
        }

        // Handle Denominator (Subtractions)
        let den_factors = Factorization::run(*den);
        for p in den_factors.factors {
            if let Some(&col) = prime_map.get(&p) {
                instr[col] -= 1;
            }
        }
        instrs.push(Instr::new(instr));
    }

    Program { instrs }
}

// Load all program strings from a file (without parsing).
pub fn load_lines(filename: &str) -> Vec<String> {
    let file = File::open(filename).expect("File not found");
    let reader = io::BufReader::new(file);

    reader
        .lines()
        .filter_map(|line| {
            let line = line.ok()?;
            let trimmed = line.trim();
            if !trimmed.is_empty() && !trimmed.starts_with('#') && !trimmed.starts_with("//") {
                Some(trimmed.to_string())
            } else {
                None
            }
        })
        .collect()
}

// Parse a filename with optional trailing :n record number (defaults to 0).
pub fn split_filename_record(filename_record: &str) -> (String, usize) {
    match filename_record.split_once(":") {
        Some((left, right)) => (left.to_string(), right.parse().expect("Invalid record_num")),
        None => (filename_record.to_string(), 0),
    }
}

pub fn load_program(input: &str) -> Option<Program> {
    // First check if it is a literal program
    if input.contains('[') {
        return Some(parse_program(input));
    }

    // Then attemp to load it as a filename/filename:n
    let (filename, record_num) = split_filename_record(input);
    let lines = load_lines(&filename);
    let prog_str = lines.iter().nth(record_num)?;
    Some(parse_program(prog_str))
}
